text/3579-pki.md

```diff
@@ -0,0 +1,262 @@
+
+The root key shall follow a `5-of-9` authentication model for all operations. We consider this a reasonable middle ground of quorum to prevent malicious activity, while allowing for rapid response to an event requiring a quorum. These events are iterated below in [When the Quorum will be needed][when-the-quorum-will-be-needed].
+
+This proposal delegates the authority of selecting and managing the quorum membership to the Rust Project's Leadership Council. We recommend they be selected from trusted individuals within the Project and Foundation. This is a position of trust, not one of authority over anything other than safe handling of key material; the 9 members are responsible for executing quorum operations, and providing transparency and trust (through their refusal to participate in any malicious operations), but not for deciding independently what key operations should happen.
+
+Authentication to the CloudHSM instance shall be limited to only a [Signing Console][signing-console] at the network layer; at the application layer, we shall create IAM roles and accounts for all members of the quorum. These accounts will required two factor authentication and have restrictive password, timeout, and security policies applied. These roles shall only be used for Quorum signing events. There will be documentation for the quorum members on best practices for handling their credentials.
+
+These individuals should be available in the event quorum is needed for root key operations. These roles can and should be reappointed as needed, but to simplify logistics, these roles should not require rotation more often than 2-3 years. (Operations requiring quorum are expected to be rare, and an annual rotation would make rotation of the quorum group the vast majority of quorum operations.)
+
+##### When the Quorum will be needed
+[when-the-quorum-will-be-needed]: #when-the-quorum-will-be-needed
+
+- Changes to quorum membership or thresholds shall require a quorum (enforced by the HSM authentication)
+- Replacement of the Root Key in the event of expiration or compromise
+- Creation, signing, and issuance of a new Delegate Key
+- Creation, signing, and issuance of a replacement Delegate Key in the event of expiration or compromise
+
+##### Quorum Threat Mitigations
+
+- To reduce the feasibility of coercion or collusion, we recommend that no more than 3 (M-2) of the 9 people be affiliated with any one company or institution, or be residents or citizens of the same country. (If in the future the quorum threshold is changed, this limit should be changed accordingly to remain M-2).
```

Comment:
Perhaps a difficult question, but do we really think this protects us against coercion or collusion? It seems to me that we're going to a great deal of trouble to protect the roots, but we're not going to anywhere near as much trouble to protect intermediate or leaf certificates. For example, I don't see any scheme that's near-term viable for requiring this kind of quorum on changes to bors or changes to promote-release infrastructure. A change there can easily sign arbitrary binaries (those pieces of infra run on network-connected instances and changes are made with 1.5 people looking at the code (+0.5 for optional code review)). Lots of code lands in rust-lang/rust with exactly one person having reviewed it, too: for example any subtree sync.

At the same time, there are also other risks: The CloudHSM instance is physically in one country, and the AWS keys are also homed in one country (they are symmetric keypairs for sigv4 so AWS also has a copy). Obviously it would be a very, uh, loud thing if we found that they were used without our quorum participating -- but fundamentally it feels not that much harder than delivering an court order to ~2 people in the US (as an example).

Reply:
@Mark-Simulacrum The signing scheme for the crates.io indexes, by way of example, will incorporate Certificate Transparency and Binary Transparency, so any signature will only be valid if publicly logged, which means it shouldn't be possible to sign a malicious update *silently* (e.g. to a subset of users).

Reply:
Perhaps (without more details on the widespread checks that CT logs have logged it's hard to be certain), but I would argue that further pushes against going to all this trouble. If we're confident people can't silently issue, then why bother protecting roots this much? A KMS key or AWS Private CA are more managed solutions and seem like they'd deliver similar value overall, presuming we're so confident in the transparency solutions.

And if we're not, then I think my earlier point stands: this scheme doesn't actually get you that much more.

Reply:
Addressed in the changes we just pushed: detailed threat model, and discussion of the alternative of just using a transparency log for everything.